esp-hal/src/timer/systimer.rs

```diff
@@ -138,150 +152,551 @@ impl<'d> SystemTimer<'d, Blocking> {
+}
+
+/// The modes of a comparator.
+#[derive(Copy, Clone)]
+pub enum ComparatorMode {
+    /// The comparator will generate interrupts periodically.
+    Period,
+
+    /// The comparator will generate an interrupt when the unit reaches the
+    /// target.
+    Target,
+}
+
+impl Unit<'static, 0> {
+    /// Conjure a system timer unit out of thin air.
+    ///
+    /// # Safety
+    ///
+    /// Users must take care to ensure that only one reference to the unit is
+    /// in scope at any given time.
+    pub const unsafe fn conjure() -> Self {
+        Self(PhantomData)
+    }
+}
+
+#[cfg(not(esp32s2))]
+impl Unit<'static, 1> {
+    /// Conjure a system timer unit out of thin air.
+    ///
+    /// # Safety
+    ///
+    /// Users must take care to ensure that only one reference to the unit is
+    /// in scope at any given time.
+    pub const unsafe fn conjure() -> Self {
+        Self(PhantomData)
     }
 }
 
-impl<T, const CHANNEL: u8> InterruptConfigurable for Alarm<T, Blocking, CHANNEL> {
+/// A unit whose value cannot be updated.
+pub struct FrozenUnit<'d, const CHANNEL: u8>(&'d Unit<'d, CHANNEL>);
+
+impl<'d, const CHANNEL: u8> FrozenUnit<'d, CHANNEL> {
+    /// Creates a frozen unit. You will no longer be allowed
+    /// direct access to this unit until all the alarms created
+    /// from the unit are dropped.
+    pub fn new(unit: &'d mut Unit<'d, CHANNEL>) -> Self {
+        Self(unit)
+    }
+
+    fn borrow(&self) -> &'d Unit<'d, CHANNEL> {
+        self.0
+    }
+}
+
+/// Alarms created from the System Timer peripheral.
+pub struct SysTimerAlarms<MODE, DM: Mode> {
+    /// Alarm 0
+    pub alarm0: Alarm<'static, MODE, DM, 0, 0>,
+    /// Alarm 1
+    pub alarm1: Alarm<'static, MODE, DM, 1, 0>,
+    /// Alarm 2
+    pub alarm2: Alarm<'static, MODE, DM, 2, 0>,
+
+    /// Unit 1
+    ///
+    /// Leftover unit which wasn't used to create the three alarms.
+    #[cfg(not(esp32s2))]
+    pub unit1: Unit<'static, 1>,
+}
+
+/// A marker for a [Alarm] in target mode.
+#[derive(Debug)]
+pub struct Target;
+
+/// A marker for a [Alarm] in periodic mode.
+#[derive(Debug)]
+pub struct Periodic;
+
+/// A single alarm.
+pub struct Alarm<'d, MODE, DM, const COMP: u8, const UNIT: u8>
+where
+    DM: Mode,
+{
+    comparator: Comparator<'d, COMP>,
+    unit: &'d Unit<'d, UNIT>,
+    config: &'d Mutex<Cell<Option<Config<'d>>>>,
+    _pd: PhantomData<(MODE, DM)>,
+}
+
+impl<'d, T, DM, const COMP: u8, const UNIT: u8> Debug for Alarm<'d, T, DM, COMP, UNIT>
+where
+    DM: Mode,
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("Alarm")
+            .field("comparator", &COMP)
+            .field("unit", &UNIT)
+            .finish()
+    }
+}
+
+impl<'d, T, const COMP: u8, const UNIT: u8> Alarm<'d, T, Blocking, COMP, UNIT> {
+    /// Creates a new alarm from a comparator and unit, in blocking mode.
+    pub fn new(
+        comparator: Comparator<'d, COMP>,
+        unit: &FrozenUnit<'d, UNIT>,
```

Comment:
I think the `FrozenUnit` idea is just replicating the `PeripheralRef` pattern, so maybe we can just use that instead? I'm not actually sure about the interaction with peripheral ref and immutable references though.

Reply:
If I've understood the `PeripheralRef` pattern correctly, it's just a glorified version of the `BorrowMut` trait. Allowing you to be generic over borrowing or owning something.

The `FrozenUnit` idea here (which I think is different) is meant to achieve two things:
One is to make the user give up direct access to the unit. If the user updates the counter value on the unit, the `Alarm`s will not work correctly and the invariant on `embassy_time_driver` about the timer never going backwards would be broken. This is why `FrozenUnit` requires a `&mut` in its constructor, the cannot touch the unit whilst it's frozen.
(As an aside, the `SystemTimer::now()` implementation also has this problem and I fear fixing it just means getting rid of it)

Two is to allow users create multiple `Alarm`s from a single unit. It'd be too restrictive for an Alarm to take full ownership of the `Unit`.
So users can freeze a unit, then use it to create as many alarms from it as necessary.

I don't think `PeripheralRef` can solve the above problems.

Reply:
Ah okay, I see the use now. You're right, it's a bit different to peripheral ref.

> (As an aside, the SystemTimer::now() implementation also has this problem and I fear fixing it just means getting rid of it)

Hmm, I don't think this would be too bad. It's only used in RISCV delay, and I think we can probably replace that with a implementation using `get_time` (https://docs.esp-rs.org/esp-hal/esp-hal/0.19.0/esp32c6/esp_hal/time/fn.current_time.html)

Reply:
Sadly `get_time` uses `SystemTimer::now()` 😄.

It's a shame RISCV seems to lack the CPU timers that Xtensa provides.